File: server/sceneWb.ts
// Scene white-balance match ---------------------------------------------------
// Shots of the same scene (a burst) come back from the AI edit with slightly
// different white balance, so near-identical frames drift in colour. We group
// the favourites by capture time and compute, per group, a shared per-channel
// gain that pulls every member's neutral onto a common target — deterministic
// and content-robust, unlike a distribution transfer. The gains are cached and
// only recomputed when the favourite set (or its files) changes.
import { spawnSync } from "node:child_process";
import { existsSync, statSync } from "node:fs";
import { createHash } from "node:crypto";
import { join } from "node:path";
import { db } from "./db.ts";
import { REPO_ROOT } from "./config.ts";

const SCENE_SCRIPT = join(REPO_ROOT, "scripts", "scene_wb.py");
// Max gap between consecutive shots to still count as the same scene. A burst
// is seconds apart; unrelated shots are minutes apart. Kept generous because
// the correction is WB-only (gentle) — over-grouping never breaks an image.
const GROUP_GAP_MS = 40_000;

type Gain = [number, number, number];

let cache: { sig: string; gains: Map<string, Gain> } | null = null;

type FavRow = { id: string; taken_at: number | null; image_path: string };

function favorites(): FavRow[] {
  return db()
    .query<FavRow, []>(
      `SELECT p.id AS id, p.taken_at AS taken_at, v.image_path AS image_path
       FROM photos p
       JOIN versions v ON v.id = p.favorite_version_id
       WHERE p.favorite_version_id IS NOT NULL
       ORDER BY (p.taken_at IS NULL) ASC, p.taken_at ASC, p.id ASC`,
    )
    .all();
}

/** Split favourites into scene groups by capture-time proximity. */
function groupByScene(rows: FavRow[]): FavRow[][] {
  const groups: FavRow[][] = [];
  let cur: FavRow[] = [];
  let prev: number | null = null;
  for (const r of rows) {
    if (!existsSync(r.image_path)) continue;
    if (
      r.taken_at == null ||
      prev == null ||
      r.taken_at - prev > GROUP_GAP_MS
    ) {
      if (cur.length) groups.push(cur);
      cur = [r];
    } else {
      cur.push(r);
    }
    prev = r.taken_at;
  }
  if (cur.length) groups.push(cur);
  return groups;
}

function signature(rows: FavRow[]): string {
  const parts = rows.map((r) => {
    const m = existsSync(r.image_path)
      ? Math.round(statSync(r.image_path).mtimeMs)
      : 0;
    return `${r.id}:${r.taken_at ?? "x"}:${m}`;
  });
  return createHash("sha1")
    .update(JSON.stringify({ gap: GROUP_GAP_MS, parts }))
    .digest("hex");
}

function compute(): Map<string, Gain> {
  const rows = favorites();
  const groups = groupByScene(rows).filter((g) => g.length >= 2);
  if (!groups.length) return new Map();
  const manifest = {
    groups: groups.map((g) => g.map((r) => ({ id: r.id, path: r.image_path }))),
  };
  const r = spawnSync("python3", [SCENE_SCRIPT], {
    input: JSON.stringify(manifest),
    encoding: "utf8",
    timeout: 120_000,
    maxBuffer: 8 * 1024 * 1024,
  });
  if (r.status !== 0 || !r.stdout) return new Map();
  try {
    const raw = JSON.parse(r.stdout) as Record<string, Gain>;
    return new Map(Object.entries(raw));
  } catch {
    return new Map();
  }
}

/** Per-photo shared WB gain, or null if the photo isn't in a multi-shot scene.
 *  Cached; recomputed only when the favourite set or its files change. */
export function wbGainFor(photoId: string): Gain | null {
  const rows = favorites();
  const sig = signature(rows);
  if (!cache || cache.sig !== sig) {
    cache = { sig, gains: compute() };
  }
  return cache.gains.get(photoId) ?? null;
}

File: scripts/scene_wb.py
#!/usr/bin/env python3
"""Scene white-balance match.

Given scene groups (shots of the same scene), compute for each shot a fixed
per-channel gain that pulls its neutral mean onto the group's shared neutral
target. Applying an identical target to every member removes the white-balance
drift that the GPT edit introduces between near-identical shots — deterministic,
content-robust (it only scales channels, never transfers content statistics).

Input  (stdin JSON): {"groups": [[{"id","path"}, ...], ...]}
Output (stdout JSON): {"<id>": [gr, gg, gb], ...}   # only for groups of >= 2

Singletons are omitted (no group ⇒ keep per-image AWB downstream).
"""
import sys
import json
import numpy as np
from PIL import Image, ImageOps

LUMA = np.array([0.2126, 0.7152, 0.0722], np.float32)
CLAMP_LO, CLAMP_HI = 0.80, 1.25  # keep the correction gentle, never extreme


def neutral_mean(path):
    """Mean RGB of the near-neutral mid-tone pixels — the WB signature."""
    im = ImageOps.exif_transpose(Image.open(path)).convert("RGB")
    im.thumbnail((520, 520))
    a = np.asarray(im, np.float32)
    luma = (a * LUMA).sum(-1)
    mx = a.max(-1)
    mn = a.min(-1)
    sat = (mx - mn) / np.clip(mx, 1, None)
    m = (luma > 45) & (luma < 215) & (sat < 0.18)
    if m.sum() < 300:
        m = (luma > 35) & (luma < 225)
    if m.sum() < 50:
        return a.reshape(-1, 3).mean(0)
    return a[m].reshape(-1, 3).mean(0)


def main():
    req = json.load(sys.stdin)
    out = {}
    for group in req.get("groups", []):
        if len(group) < 2:
            continue
        means = {}
        for item in group:
            try:
                means[item["id"]] = neutral_mean(item["path"])
            except Exception:
                pass
        if len(means) < 2:
            continue
        target = np.mean(list(means.values()), axis=0)  # shared neutral target
        for pid, nm in means.items():
            gain = np.clip(target / np.clip(nm, 1, None), CLAMP_LO, CLAMP_HI)
            out[pid] = [round(float(x), 4) for x in gain]
    json.dump(out, sys.stdout)


if __name__ == "__main__":
    main()
